File: ios/HeatherWidget/MoonPhase.swift
import Foundation

enum MoonPhase: String {
    case newMoon = "New"
    case waxingCrescent = "Wax Cres"
    case firstQuarter = "1st Qtr"
    case waxingGibbous = "Wax Gib"
    case fullMoon = "Full"
    case waningGibbous = "Wan Gib"
    case thirdQuarter = "3rd Qtr"
    case waningCrescent = "Wan Cres"

    var sfSymbol: String {
        switch self {
        case .newMoon: return "moonphase.new.moon"
        case .waxingCrescent: return "moonphase.waxing.crescent"
        case .firstQuarter: return "moonphase.first.quarter"
        case .waxingGibbous: return "moonphase.waxing.gibbous"
        case .fullMoon: return "moonphase.full.moon"
        case .waningGibbous: return "moonphase.waning.gibbous"
        case .thirdQuarter: return "moonphase.last.quarter"
        case .waningCrescent: return "moonphase.waning.crescent"
        }
    }
}

/// Synodic month length in days.
private let synodicMonth = 29.53058770576

/// Reference new moon: Jan 18, 2026 19:51 UTC.
private let referenceNewMoon: Date = {
    var c = DateComponents()
    c.year = 2026; c.month = 1; c.day = 18
    c.hour = 19; c.minute = 51
    c.timeZone = TimeZone(identifier: "UTC")
    return Calendar.current.date(from: c)!
}()

func moonAge(for date: Date = Date()) -> Double {
    let seconds = date.timeIntervalSince(referenceNewMoon)
    let days = seconds / 86400
    let age = days.truncatingRemainder(dividingBy: synodicMonth)
    return age < 0 ? age + synodicMonth : age
}

func getMoonPhase(for date: Date = Date()) -> MoonPhase {
    let fraction = moonAge(for: date) / synodicMonth

    if fraction < 0.04 { return .newMoon }
    if fraction < 0.21 { return .waxingCrescent }
    if fraction < 0.29 { return .firstQuarter }
    if fraction < 0.46 { return .waxingGibbous }
    if fraction < 0.54 { return .fullMoon }
    if fraction < 0.71 { return .waningGibbous }
    if fraction < 0.79 { return .thirdQuarter }
    if fraction < 0.96 { return .waningCrescent }
    return .newMoon
}

func moonIllumination(for date: Date = Date()) -> Int {
    let age = moonAge(for: date)
    let fraction = age / synodicMonth
    let illum = (1 - cos(2 * .pi * fraction)) / 2 * 100
    return Int(illum.rounded())
}

File: android/app/src/main/kotlin/com/heather/heather/widget/MoonPhase.kt
package com.totms.heather.widget

import com.totms.heather.R
import java.util.Calendar
import java.util.TimeZone
import kotlin.math.cos
import kotlin.math.roundToInt

enum class MoonPhaseType(val label: String, val iconRes: Int) {
    NEW_MOON("New", R.drawable.ic_moon_new),
    WAXING_CRESCENT("Wax Cres", R.drawable.ic_moon_waxing_crescent),
    FIRST_QUARTER("1st Qtr", R.drawable.ic_moon_first_quarter),
    WAXING_GIBBOUS("Wax Gib", R.drawable.ic_moon_waxing_gibbous),
    FULL_MOON("Full", R.drawable.ic_moon_full),
    WANING_GIBBOUS("Wan Gib", R.drawable.ic_moon_waning_gibbous),
    LAST_QUARTER("3rd Qtr", R.drawable.ic_moon_last_quarter),
    WANING_CRESCENT("Wan Cres", R.drawable.ic_moon_waning_crescent);
}

private const val SYNODIC_MONTH = 29.53058770576

/** Reference new moon: Jan 18, 2026 19:51 UTC */
private val referenceNewMoonMillis: Long by lazy {
    val cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
    cal.set(2026, Calendar.JANUARY, 18, 19, 51, 0)
    cal.set(Calendar.MILLISECOND, 0)
    cal.timeInMillis
}

fun moonAge(timeMillis: Long = System.currentTimeMillis()): Double {
    val days = (timeMillis - referenceNewMoonMillis) / 86_400_000.0
    val age = days.mod(SYNODIC_MONTH)
    return if (age < 0) age + SYNODIC_MONTH else age
}

fun getMoonPhase(timeMillis: Long = System.currentTimeMillis()): MoonPhaseType {
    val fraction = moonAge(timeMillis) / SYNODIC_MONTH
    return when {
        fraction < 0.04 -> MoonPhaseType.NEW_MOON
        fraction < 0.21 -> MoonPhaseType.WAXING_CRESCENT
        fraction < 0.29 -> MoonPhaseType.FIRST_QUARTER
        fraction < 0.46 -> MoonPhaseType.WAXING_GIBBOUS
        fraction < 0.54 -> MoonPhaseType.FULL_MOON
        fraction < 0.71 -> MoonPhaseType.WANING_GIBBOUS
        fraction < 0.79 -> MoonPhaseType.LAST_QUARTER
        fraction < 0.96 -> MoonPhaseType.WANING_CRESCENT
        else -> MoonPhaseType.NEW_MOON
    }
}

fun moonIllumination(timeMillis: Long = System.currentTimeMillis()): Int {
    val age = moonAge(timeMillis)
    val fraction = age / SYNODIC_MONTH
    val illum = (1 - cos(2 * Math.PI * fraction)) / 2.0 * 100.0
    return illum.roundToInt()
}
